### PR TITLE
Bare metal cluster / MetalLB fixups

### PR DESCRIPTION
**Pull request by logan2211**
Several fixes to calico-node templates to make running calico on bare metal clusters smooth.
* Don't advertise prefixes to mesh peers when tunnelling is disabled on an IPPool
* Setting `passive` on localhost peers 127.0.0.1 and ::1 [so MetalLB can peer with calico-node correctly](https://github.com/metallb/metallb/issues/114).
* Support a passthru community tag for MetalLB routes which will ferry peer announcements from downstream peers to global/node peers, but not to the tunnel mesh peers. This makes `externalTrafficPolicy` work as expected in both modes.
* Only force peer source address on mesh peers. Source address for upstream peers will be the originating interface of the BGP connection per the node's routing table / PBR rules.

### CODE REVIEW

**Comment by caseydavenport** [comment]
@logan2211 thanks for the PR and sorry for the delay in review. I will try to pick this up next week.

**Comment by ivan046** [comment]
@logan2211 Could you, please, find some time to fix that?

**Comment by caseydavenport** [comment]
> Don't advertise prefixes to mesh peers when tunnelling is disabled on an IPPool

I'm still thinking through some of the changes in this PR, but I had a question about this one specifically. Could you provide some more information on why this is desired? My thought so far is that this would break deployments that use full-mesh mode without encap, since mesh peers won't learn routes from each other. Or am I misunderstanding?

**Comment by logan2211** [comment]
> > Don't advertise prefixes to mesh peers when tunnelling is disabled on an IPPool
> 
> Could you provide some more information on why this is desired? My thought so far is that this would break deployments that use full-mesh mode without encap, since mesh peers won't learn routes from each other. Or am I misunderstanding?

Sure, the routes are still announced to global / node peers. So in a situation where no encap/mesh is used (ie ipip and vxlan are disabled), the routes are announced to the upstream network and that network is expected to route traffic accordingly. In situations where you may have some ippools which are routed by the ipip/vxlan overlay, and other ippools that are routed by the upstream network, this fixes the latter scenario.

**Comment by logan2211** [comment]
An example ippol where this would be useful would be like:
```
apiVersion: projectcalico.org/v3
kind: IPPool
metadata:
  name: external.ippool-1
spec:
  cidr: 10.199.0.0/24
  ipipMode: Never
  vxlanMode: Never
```

In this case, with 7062d782ebb2bff5946bbed8643ee2f0294ad360 applied, calico will no longer announce the routes over the BGP sessions it establishes with other nodes, but it will still announce them to global/node BGPPeers.

**Comment by caseydavenport** [comment]
> In this case, with 7062d78 applied, calico will no longer announce the routes over the BGP sessions it establishes with other nodes, but it will still announce them to global/node BGPPeers.

I don't think we can make that change the way it is in this PR. That IP pool might represent a cluster CIDR for a cluster that is running in an L2 network, where no encap is needed between the nodes. In that case, we still need BGP to announce routes to other nodes in the cluster, even though the traffic won't be encapsulated.

I think we should pull this bit out of the current PR and discuss requirements / design in another issue.

File: etc/calico/confd/templates/bird_ipam.cfg.template [hunk 1]
```diff
@@ -1,5 +1,37 @@
 # Generated by confd
 filter calico_export_to_bgp_peers {
+  if bgp_passthru() then accept;
+  calico_aggr();
+{{- $static_key := "/staticroutes"}}
+{{- if ls $static_key}}
+
+  # Export static routes.
```
Comment (caseydavenport) [inline etc/calico/confd/templates/bird_ipam.cfg.template:8]: Looks like this logic is duplicated now in export_to_bgp_peers and export_to_bgp_mesh. Probably worth using a common function - e.g., `if static_route_accept() then accept;` WDYT?

**Comment by caseydavenport** [comment]
Just wanted to summarize where I think we are with this PR:

> Don't advertise prefixes to mesh peers when tunnelling is disabled on an IPPool

I don't think this is something we can accept right now - there are scenarios currently that rely on this behavior. Described in [this comment](https://github.com/projectcalico/confd/pull/313/#issuecomment-602694901).

> Setting passive on localhost peers 127.0.0.1 and ::1 so MetalLB can peer with calico-node correctly.

This seems reasonable to me 👍 

> Support a passthru community tag for MetalLB routes which will ferry peer announcements from downstream peers to global/node peers, but not to the tunnel mesh peers. This makes externalTrafficPolicy work as expected in both modes.

I''m not 100% sure I understand this change (sorry if I'm being dull). It seems to me that this change lets us specify a community, where any matching prefixes advertised from MetalLB will automatically accepted by our export filter. Is that the intended motivation for this? 

If it is, then I think we already have a mechansim within Calico to do this - namely, by creating an IP pool that contains the prefixes. I think we would want to instruct users to do that rather than add a new environment variable to manage this. Am I missing something?

> Only force peer source address on mesh peers. Source address for upstream peers will be the originating interface of the BGP connection per the node's routing table / PBR rules.

I think this change also makes sense. 

Aside from the above, I think the CLA bot is still waiting on a signed CLA before we can accept any of these changes. 

Appreciate your effort and patience on this one! I'd really like to fix up the MetalLB integration pain points.

**Comment by logan2211** [comment]
Thanks for the reviews @caseydavenport and @neiljerram.. apologies for the delayed responses!

> I don't think we can make that change the way it is in this PR. That IP pool might represent a cluster CIDR for a cluster that is running in an L2 network, where no encap is needed between the nodes. In that case, we still need BGP to announce routes to other nodes in the cluster, even though the traffic won't be encapsulated.

> I think we should pull this bit out of the current PR and discuss requirements / design in another issue.

Ok I think I understand what you're saying about the layer2 use case. Maybe this would be a setting on the IPPool allowing calico to send the routes to the global/node peers but not mesh peers? For a bare-metal cluster build out, this bug affects me badly.

I was really surprised to see calico announcing these prefixes over the tunnel mesh when tunnelling was not enabled, and of course it totally breaks the routing for that IP pool, because every node is announcing the prefix yet only 1 is prepared to handle ingress.

Maybe an attribute on IPPool.spec `meshRouting: true/false` with true as the default?

> I''m not 100% sure I understand this change (sorry if I'm being dull). It seems to me that this change lets us specify a community, where any matching prefixes advertised from MetalLB will automatically accepted by our export filter. Is that the intended motivation for this?

> If it is, then I think we already have a mechansim within Calico to do this - namely, by creating an IP pool that contains the prefixes. I think we would want to instruct users to do that rather than add a new environment variable to manage this. Am I missing something?

Creating an IPPool doesn't work correctly for this. Yes, Calico will accept the announcements from MetalLB once an IPPool is created. Unfortunately, it will then take those announcements and share them to mesh peers, and then the mesh peers will share those announcements with upstreams from each node, completely defeating the point of `externalTrafficPolicy: Local`. 3d3dfc3e8f22edcf1c638e62a0c42898df243c43 aims to fix that by allowing a community tag that bypasses the mesh for certain routes.

MetalLB uses the community regardless of externalTrafficPolicy, it will either:
1) Announce the route to calico from each node in the daemonset when `externalTrafficPolicy: Cluster`
 is set, and calico will pass that announcement to upstreams.
2) Announce the route only from nodes hosting service endpoints as in the case of `externalTrafficPolicy: Local`, calico will pass the announcements to upstreams from these nodes only.

> Aside from the above, I think the CLA bot is still waiting on a signed CLA before we can accept any of these changes.

Ok, will look into it.

> As a general principle, when there are N conceptually separable fixes, please can we have N separate PRs? It will make it so much easier to hold any discussion we need on each point, and for others in future to refer back to and understand that. And to modify or even revert one of the fixes, in case that becomes necessary later on.

> I think that splitting is even worth doing now with this PR, given that there are different ongoing discussions for the different proposed fixes.

I will chop this up into separate PRs.

> Re passive, I like the idea of going a bit further and adding a BGPPeer field to control this, as per projectcalico/calico#1603

I don't disagree, and as a user who needs this to work, I would have just used that if it existed. But alas the issue has been open for 2+ years without a PR. I hope you will consider accepting my proposed fix in the interim, even if it is not ideal, and then it can be removed if/when someone takes up the work in projectcalico/calico#1603.

File: etc/calico/confd/templates/bird_ipam.cfg.template (hunk 1 above)
Comment (logan2211) [reply etc/calico/confd/templates/bird_ipam.cfg.template:8]: Sure, will do.

**Pull request closed without merging**